capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+                        // wait until ready and then read data block, then done
+                        self.state.set(State::WaitReadBlocks { count: count });
+                        self.read_bytes(write_buffer, read_buffer, 1);
+                    });
+                });
+
+                self.alarm_state.set(AlarmState::Idle);
+            }
+
+            AlarmState::WaitForWriteBusy => {
+                // buffers must be available to use
+                if self.txbuffer.is_none() {
+                    panic!("No txbuffer available for timer");
+                }
+                if self.rxbuffer.is_none() {
+                    panic!("No rxbuffer available for timer");
+                }
+
+                // check card initialization again
+                self.txbuffer.take().map(|write_buffer| {
+                    self.rxbuffer.take().map(move |read_buffer| {
+                        // check if sd card is busy
+                        self.state.set(State::WaitWriteBlockBusy);
+                        self.read_bytes(write_buffer, read_buffer, 1);
+                    });
+                });
+
+                self.alarm_state.set(AlarmState::Idle);
+            }
+
+            AlarmState::Idle => {
+                // receiving an event from Idle means something was killed
+                // do nothing
+            }
+        }
+    }
+
+    pub fn set_client<C: SDCardClient>(&self, client: &'static C) {
+        self.client.set(Some(client));
+    }
+
+    pub fn is_installed(&self) -> bool {
+        // if there is no detect pin, assume an sd card is installed
+        self.detect_pin.get().map_or(true, |pin| {
+            // sd card detection pin is active low
+            pin.read() == false
+        })
+    }
+
+    pub fn is_initialized(&self) -> bool {
+        self.is_initialized.get()
+    }
+
+    // watches SD card detect pin for changes, sends callback on change
+    pub fn detect_changes(&self) {
+        self.detect_pin
+            .get()
+            .map(|pin| { pin.enable_interrupt(0, hil::gpio::InterruptMode::EitherEdge); });
+    }
+
+    pub fn initialize(&self) -> ReturnCode {
+        // if not already, set card to uninitialized again
+        self.is_initialized.set(false);
+
+        // no point in initializing if the card is not installed
+        if self.is_installed() {
+            // reset the SD card in order to start initializing it
+            self.txbuffer.take().map_or(ReturnCode::ENOMEM, |txbuffer| {
+                self.rxbuffer.take().map_or(ReturnCode::ENOMEM, move |rxbuffer| {
+                    self.state.set(State::InitReset);
+                    self.send_command(SDCmd::CMD0, 0x0, txbuffer, rxbuffer, 10);
+
+                    // command started successfully
+                    ReturnCode::SUCCESS
+                })
+            })
+        } else {
+            // no sd card installed
+            ReturnCode::EOFF
```

Comment:
I don't love EOFF here. Implies you could make a call to turn something on and then it would work. Maybe we need a new return code, maybe just FAIL